node/core/dispute-coordinator/src/scraping/mod.rs

```diff
@@ -58,18 +60,43 @@ const LRU_OBSERVED_BLOCKS_CAPACITY: NonZeroUsize = match NonZeroUsize::new(20) {
 ///
 /// Concretely:
 ///
-/// - Monitors for inclusion events to keep track of candidates that have been included on chains.
+/// - Monitors for `CandidateIncluded` events to keep track of candidates that have been included on chains.
+/// - Monitors for `CandidateBacked` events to keep track of all backed candidates.
 /// - Calls `FetchOnChainVotes` for each block to gather potentially missed votes from chain.
 ///
 /// With this information it provides a `CandidateComparator` and as a return value of
 /// `process_active_leaves_update` any scraped votes.
 pub struct ChainScraper {
 	/// All candidates we have seen included, which not yet have been finalized.
-	included_candidates: HashSet<CandidateHash>,
+	included_candidates: candidates::RefCountedCandidates,
+	/// All candidates we have seen backed
+	backed_candidates: candidates::RefCountedCandidates,
 	/// including block -> `CandidateHash`
 	///
 	/// We need this to clean up `included_candidates` on finalization.
-	candidates_by_block_number: BTreeMap<BlockNumber, HashSet<CandidateHash>>,
+	included_candidates_by_block_number: BTreeMap<BlockNumber, HashSet<CandidateHash>>,
+	/// Keeps track at which block a `CandidateHash` was backed. If a candidate was backed
+	/// and included - it will be removed once the block at its height is finalized. However
+	/// a candidate can be backed and never included. To avoid leaking memory in such cases
+	/// we keep track at which block height the candidate was backed. We remove backed (and
+	/// not included) candidates when `BACKED_CANDIDATE_MAX_LIFETIME` more blocks are
+	/// finalized after the block they were initially backed in.
+	/// E.g. if `BACKED_CANDIDATE_MAX_LIFETIME = 2` when block 4 is finalized we will remove
+	/// all candidates backed in block 2 which were not already cleaned up on finalization.
+	///
+	/// Please note that if a candidate is backed AND included it will be cleaned up on
+	/// finalization thanks to the tracking in `included_candidates_by_block_number`.
+	///
+	/// Also please note that this `BTreeMap` keeps information when a candidate was backed
```

Comment:
I'm hesitant about this. It's not perfect but fixing it will require keeping each backed `CandidateHash` in three separate data structures.
@eskimor @BradleyOlson64 -> what's your opinion?

Reply:
One potential option could be a bijective map I suppose.

Reply:
Bijective map will definitely help if we decide the storage overhead is worth it.

[bimap](https://crates.io/crates/bimap) seems to be popular, despite its last release is 10 months ago. As far as I can see internally it uses two `HashMap`s - [link](https://github.com/billyrieger/bimap-rs/blob/0a84fb7620b1d8542c3a52dc6b05ddf532d0a7cb/src/hash.rs#L21).

My usecase here is very simple so we can stitch two `HashMap`s together and avoid the extra dependency.

Reply:
The longer I think about this the more convinced I am that `backed_candidates_by_block_number` should be cleaned properly.

Reply:
This actually is not relevant because backed and included candidates should be treated in the same way. So I don't have to remove one before the other and vice versa.